prowler/lib/scan/scan.py

```diff
@@ -1,60 +1,155 @@
+                    check_findings = execute(
+                        service,
+                        check_name,
+                        self._provider,
+                        custom_checks_metadata,
+                    )
+
+                    # Store findings
+                    self._findings.extend(check_findings)
+
+                    # Remove the executed check
+                    self._service_checks_to_execute[service].remove(check_name)
+                    if len(self._service_checks_to_execute[service]) == 0:
+                        self._service_checks_to_execute.pop(service, None)
+                    # Add the completed check
+                    if service not in self._service_checks_completed:
+                        self._service_checks_completed[service] = set()
+                    self._service_checks_completed[service].add(check_name)
+                    self._number_of_checks_completed += 1
+
+                    # This should be done just once all the service's checks are completed
+                    # This metadata needs to get to the services not within the provider
+                    # since it is present in the Scan class
+                    self._provider.audit_metadata = update_audit_metadata(
+                        self._provider.audit_metadata,
+                        self.get_completed_services(),
+                        self.get_completed_checks(),
+                    )
+
+                    # Yield the progress and the findings
+                    yield self._provider.audit_metadata, check_findings
+
+                # If check does not exists in the provider or is from another provider
+                except ModuleNotFoundError:
+                    logger.error(
+                        f"Check '{check_name}' was not found for the {self._provider.type.upper()} provider"
+                    )
+                except Exception as error:
+                    logger.error(
+                        f"{check_name} - {error.__class__.__name__}[{error.__traceback__.tb_lineno}]: {error}"
+                    )
+        except Exception as error:
+            logger.error(
+                f"{check_name} - {error.__class__.__name__}[{error.__traceback__.tb_lineno}]: {error}"
+            )
+
+    def get_completed_services(self):
+        return self._service_checks_completed.keys()
+
+    def get_completed_checks(self):
+        completed_checks = set()
+        for checks in self._service_checks_completed.values():
+            completed_checks.update(checks)
+        return completed_checks
+
+
+def get_service_name_from_check_name(check_name: str) -> str:
+    """
+    get_service_name_from_check_name returns the service name for a given check name.
+
+    Example:
+        get_service_name_from_check_name("ec2_instance_public") -> "ec2"
+    """
+    return check_name.split("_")[0]
+
+
+def get_service_checks_to_execute(checks_to_execute: set[str]) -> dict[str, set[str]]:
```

Comment:
Please add docstings for all the functions and methods.

Reply:
Done